# The Monty Hall Problem with N Doors
## Probability, Strategy Comparison and Simulation for Multiple Doors

## Abstract

This project considers the Monty Halll problem for a varying number of doors using both theoretical reasoning and simulation in Python. The aim is to see how the probabilities change and how different strategies perform.

The results show that switching is consistently the better choice, and its advantage becomes more noticeable as the number of doors increases. The simulated results are close to the theoretical values, which confirms that the model behaves as expected.

## Introduction
The Monty Hall problem is a well-known example in probability theory, based on a simple game involving multiple doors. Behind one of the doors there is a car, while behind the rest there are goats.

The player first chooses a door. Then the host, who knows where the car is, opens some of the remaining doors and reveals goats. After that, the player has a choice: to stay with the original selection or to switch to another unopened door.

In many cases, it is assumed that both options lead to the same probability of winning. However, this is not true, and a deeper analysis shows that one strategy is clearly better.

In this project, the problem is analyzed for a larger number of doors. The goal is to understand how the probabilities change and to compare different strategies using both theoretical reasoning and simulations in Python.

## Methods

### Simulation logic

The problem is extended to \(N\) doors instead of only three. One door contains a car, while the others contain goats.
At the beginning, the player chooses one door. Then the host opens all other doors that do not contain the car, leaving only one additional unopened door.

Doors that do not contain the car are implicitly treated as goats. The host always reveals such doors, ensuring that the car is never shown. This plays a key role in the game, as it provides additional information to the player.

The simulation follows these steps:

1. A car is randomly placed behind one of the doors  
2. The player randomly selects a door  
3. The host opens all doors with goats except one  
4. Two doors remain closed  
5. The player either stays or switches  

### Strategies

Three strategies are used in the simulation:

- Stay – the player keeps the initial choice  
- Switch – the player changes to the other door  
- Random – the player randomly decides

### Theoretical probabilities

For a game with \(N\) doors, the probability of winning by switching is:

$$
P(\text{switch}) = \frac{N - 1}{N}
$$

while the probability of winning by staying is:

$$
P(\text{stay}) = \frac{1}{N}
$$

### Experiments

The simulation is repeated many times in order to estimate probabilities.

Different values of \(N\) are tested, such as:
- 3 doors  
- 10 doors  
- 100 doors  

## Implementation and Results

The following code implements the simulation of the Monty Hall problem for a variable number of doors. Different strategies are tested in order to estimate their probability of winning.

The random function is used to simulate `random` choices in the game.

In [1]:
import random

The function below simulates one round of the Monty Hall game.

In [6]:
def play_monty_hall_n_doors(n=3, strategy="stay"):
    doors = list(range(n))
    
    car = random.choice(doors)
    player_choice = random.choice(doors)

    if player_choice == car:
        remaining = random.choice([d for d in doors if d != player_choice])
    else:
        remaining = car

    if strategy == "switch":
        player_choice = remaining
    elif strategy == "random":
        if random.choice([True, False]):
            player_choice = remaining

    return player_choice == car


The function `play_monty_hall_n_doors` simulates a single round of the game.

It creates the set of doors, randomly places the car, and generates the player’s initial choice. Then, depending on that choice, it determines which door remains after the host reveals the others.

Based on the selected strategy, the player either keeps the initial door, switches to the other one, or makes a random decision.

At the end, the function returns `True` if the final choice is the correct door (the player wins the car), and `False` otherwise.

Some example runs of the function are shown below:

In [3]:
print(play_monty_hall_n_doors(3, "stay"))
print(play_monty_hall_n_doors(3, "switch"))

False
True


The function `simulate_n_doors` runs the game many times in order to estimate the probability of winning.

For each trial, the function calls `play_monty_hall_n_doors` for each strategy and counts how often the player wins.

At the end, the results are converted into probabilities by dividing the number of wins by the total number of trials.

The function returns three values corresponding to the success rate of each strategy.

In [4]:
def simulate_n_doors(n=3, trials=10000):
    stay = 0
    switch = 0
    random_strategy = 0

    for _ in range(trials):
        if play_monty_hall_n_doors(n, "stay"):
            stay += 1
        if play_monty_hall_n_doors(n, "switch"):
            switch += 1
        if play_monty_hall_n_doors(n, "random"):
            random_strategy += 1

    return stay/trials, switch/trials, random_strategy/trials

In [5]:
print(simulate_n_doors(3))

(0.3401, 0.6635, 0.4946)


The results confirm that switching ($\approx 0.664 \approx \frac{2}{3}$) leads to a higher probability of winning compared to staying ($\approx 0.340 \approx \frac{1}{3}$), while the random strategy is close to $\approx 0.495 \approx \frac{1}{2}$.

The simulation estimates the probability of winning using the empirical formula:

$$
P \approx \frac{\text{number of wins}}{\text{number of trials}}
$$

In [7]:
print(simulate_n_doors(10))
print(simulate_n_doors(100))

(0.0954, 0.8997, 0.496)
(0.0094, 0.9888, 0.498)


As the number of doors increases, the advantage of switching becomes more pronounced, while staying becomes increasingly less effective.

## Discussion

The results clearly show that switching is the better strategy, and this effect becomes stronger as the number of doors increases. When there are many doors, staying almost never leads to winning.

The small differences between the simulated results and the theoretical values are due to randomness, since the simulation is based on a finite number of trials.

Further improvements could include running more simulations for higher accuracy or visualizing how the probabilities change as the number of doors increases.

## Conclusion

The results confirm that switching is the better choice in the Monty Hall problem. As the number of doors increases, this advantage becomes much more noticeable.

The simulation results closely follow the theoretical expectations, showing that the implementation works as intended.

## References

- Wikipedia: https://en.wikipedia.org/wiki/Monty_Hall_problem
- YouTube: https://www.youtube.com/watch?v=mhlc7peGlGg